onelogin_aws_cli/__init__.py

```diff
@@ -84,38 +88,78 @@ def get_saml_assertion(self):
-        if callback:
-            devices = res[0]["devices"]
-            device_id = None
-            if len(devices) > 1:
-                for i in range(0, len(devices)):
-                    print("{}. {}".format(i+1, devices[i]["device_type"]))
-                device_num = input("Which OTP Device? ")
-                device_id = devices[int(device_num)-1]["device_id"]
-            else:
-                device_id = devices[0]["device_id"]
-
-            otp_token = input("OTP Token: ")
-
-            params = {
-                "app_id": self.config["aws_app_id"],
-                "device_id": str(device_id),
-                "state_token": state_token,
-                "otp_token": otp_token
-            }
-            res = self.request("api/1/saml_assertion/verify_factor",
-                               headers, params)
-            self.saml = res
-
-    def assume_role(self):
+        if isinstance(res, list):
+            callback = res[0]["callback_url"]
+            state_token = res[0]["state_token"]
+            if callback:
+                devices = res[0]["devices"]
+                device_id = None
+                if len(devices) > 1:
+                    for i in range(0, len(devices)):
+                        print("{}. {}".format(i+1, devices[i]["device_type"]))
+                    device_num = input("Which OTP Device? ")
+                    device_id = devices[int(device_num)-1]["device_id"]
+                else:
+                    device_id = devices[0]["device_id"]
+
+                otp_token = input("OTP Token: ")
+
+                params = {
+                    "app_id": self.config["aws_app_id"],
+                    "device_id": str(device_id),
+                    "state_token": state_token,
+                    "otp_token": otp_token
+                }
+                res = self.request("api/1/saml_assertion/verify_factor",
+                                   headers, params)
+        self.saml = res
+
+    def get_arns(self):
         if not self.saml:
             self.get_saml_assertion()
+        # Parse the returned assertion and extract the authorized roles
+        aws_roles = []
+        root = ET.fromstring(base64.b64decode(self.saml))
+
+        for saml2attribute in root.iter('{urn:oasis:names:tc:SAML:2.0:assertion}Attribute'):
+            if (saml2attribute.get('Name') == 'https://aws.amazon.com/SAML/Attributes/Role'):
+                for saml2attributevalue in saml2attribute.iter('{urn:oasis:names:tc:SAML:2.0:assertion}AttributeValue'):
+                    aws_roles.append(saml2attributevalue.text)
+
+        # Note the format of the attribute value should be role_arn,principal_arn
+        # but lots of blogs list it as principal_arn,role_arn so let's reverse
+        # them if needed
+        aws_roles = [role.split(",") for role in aws_roles]
+        aws_roles = [(role, principal ) for role, principal in aws_roles]
+        self.all_roles = aws_roles
+
+    def get_role(self):
```

Comment:
You're not currently handling the case where there is one role (self.principal_arn and self.role_arn aren't getting set).

Here's a working version:
```python
    def get_role(self):
        if not self.all_roles:
            self.get_arns()

        if not self.all_roles:
            raise Exception("No roles found")

        selected_role = None

        # If I have more than one role, ask the user which one they want,
        # otherwise just proceed
        if len(self.all_roles) > 1:
            for role, principal in self.all_roles:
                print("[{}] {}".format(role, principal))
            while not selected_role:
                choice = int(input("Role : "))
                if choice in range(len(self.all_roles)):
                    selected_role = choice
                else:
                    print("Invalid role index, please try again")
        else:
            selected_role = 0

        self.role_arn, self.principal_arn = self.all_roles[selected_role]
```